opensquirrel/parser/libqasm/parser.py

```diff
@@ -96,22 +105,64 @@ def _get_bits(
             ret = [Bit(index) for index in indices]
         return ret
 
+    @classmethod
+    def _get_gate_operands(
+        cls, instruction: cqasm.semantic.GateInstruction, register_manager: RegisterManager
+    ) -> list[list[Any]]:
+        """Get the list of lists of operands of a gate.
+        Notice that a gate just has a list of operands. The outer list is needed to support SGMQ.
+        For example, for CNOT q[0, 1] q[2, 3], this function returns [[Qubit(0), Qubit(1)], [Qubit(2), Qubit(3)]].
+        """
+        ret: list[list[Any]] = []
+        for operand in instruction.operands:
+            if cls._is_qubit_type(operand):
+                ret.append(cls._get_qubits(operand, register_manager))
+            else:
+                msg = "argument is not of qubit type"
+                raise TypeError(msg)
+        return ret
+
+    @classmethod
+    def _get_named_gate_param(cls, gate: cqasm.semantic.Gate) -> Any:
+        """Get the parameter of a named gate.
+        Notice the input gate can be a composition of gate modifiers acting on a named gate.
+        """
+        if gate.name in ["inv", "pow", "ctrl"]:
+            return cls._get_named_gate_param(gate.gate)
+        return cls._ast_literal_to_ir_literal(gate.parameter)
+
+    @classmethod
+    def _get_expanded_gate_args(
+        cls, instruction: cqasm.semantic.GateInstruction, register_manager: RegisterManager
+    ) -> zip[tuple[Any, ...]]:
+        """Construct a list with a list of qubits and a list of parameters, then return a zip of both lists.
+        For example, for CRk(2) q[0, 1] q[2, 3], this function first constructs the list with a list of qubits
+        [[Qubit(0), Qubit(1)], [Qubit(2), Qubit(3)]], then appends the list of parameters [Int(2), Int(2)],
+        and finally zips the whole list and returns [(Qubit(0), Qubit(1), Int(2)), (Qubit(2), Qubit(3), Int(2))]
+        """
+        ret = cls._get_gate_operands(instruction, register_manager)
+        gate_parameter = cls._get_named_gate_param(instruction.gate)
+        if gate_parameter is not None:
+            number_of_operands = len(ret[0])
+            ret.append([gate_parameter] * number_of_operands)
+        return zip(*ret)
+
     @classmethod
     def _get_expanded_measure_args(cls, ast_args: Any, register_manager: RegisterManager) -> zip[tuple[Any, ...]]:
         """Construct a list with a list of bits and a list of qubits, then return a zip of both lists.
         For example: [(Qubit(0), Bit(0)), (Qubit(1), Bit(1))]
 
-        Notice the  list is walked in reverse mode.
+        Notice the list is walked in reverse mode.
```

Comment:
Is this best to leave as a comment above the for loop instead of a docstring?

Reply:
Good point.